### ISSUE TITLE
[Tests] Implement test utility functions and fixtures

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2734]
## Summary
Implement test infrastructure components.

## Components Needed
- Test utility functions (utils.mojo)
- Test fixtures (fixtures.mojo)
- Test configuration (conftest.mojo)
- Integration test execution logic
- TOML config loading for multi-precision tests

## Files Affected
- \`tests/helpers/utils.mojo\` (line 29)
- \`tests/helpers/fixtures.mojo\` (line 30)
- \`tests/conftest.mojo\` (line 127)
- \`tests/shared/integration/__init__.mojo\` (line 31)
- \`tests/shared/integration/test_multi_precision_training.mojo\` (line 32)

## Related TODO Lines in todo.log
- Lines 29-32: test helper/fixture TODOs
- Line 127: conftest.mojo

## Notes
Some fixtures overlap with Issue #2715 (FIXME-related fixtures).

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Based on my comprehensive analysis of the codebase, here is the detailed implementation plan for Issue #2734:

---

# Implementation Plan: Issue #2734 - Test Utility Functions and Fixtures

## Summary

This implementation plan addresses Issue #2734 to implement test infrastructure components across 5 files. The work involves completing placeholder test utilities, fixtures, and configuration in the `tests/` directory while integrating with the existing comprehensive `shared/testing/` infrastructure.

**Key Insight**: Most test infrastructure already exists in `shared/testing/` (assertions, fixtures, models, gradient checking, layer testers, dtype utilities). This issue focuses on completing the **thin wrappers** in `tests/helpers/` and adding **test execution logic** for integration tests.

**Overlap with #2715**: Issue #2715 tracks FIXME items for placeholder fixtures that depend on ExTensor maturity. This issue (#2734) tracks the general TODO items. The fixtures overlap (lines 6-14 in fixtures.mojo) should be coordinated - implement foundations here, defer ExTensor-dependent features to #2715.

## Step-by-Step Implementation Tasks

### Phase 1: Implement Test Utility Functions (tests/helpers/utils.mojo)

**File**: `/home/mvillmow/ml-odyssey/tests/helpers/utils.mojo`
**Current State**: Empty placeholder with TODO comments (lines 6-11)

**Tasks**:

1. **Implement `print_tensor(tensor: ExTensor) -> String`**
   - Pretty-print tensor with shape, dtype, and values
   - Truncate large tensors (show first/last 3 elements)
   - Format: `Tensor(shape=[2,3], dtype=float32):\n  [[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]`

2. **Implement `tensor_summary(tensor: ExTensor) -> String`**
   - Print shape, dtype, numel
   - Calculate and print min/max/mean/std statistics
   - Useful for debugging during test development

3. **Implement `compare_tensors(a: ExTensor, b: ExTensor) -> String`**
   - Detailed comparison showing where tensors differ
   - Return summary: "Shapes match: ✓, DTypes match: ✓, Max diff: 0.001 at index 5"
   - Use for debugging assertion failures

4. **Implement `benchmark[func: fn() raises -> None](iterations: Int) -> Float64`**
   - Simple performance testing helper using parametric function
   - Run function `iterations` times
   - Return average execution time in milliseconds

**Dependencies**: `shared.core.extensor`, `time` module

### Phase 2: Implement Test Fixtures (tests/helpers/fixtures.mojo)

**File**: `/home/mvillmow/ml-odyssey/tests/helpers/fixtures.mojo`
**Current State**: Empty placeholder with FIXME(#2715) and TODO(#2734) comments

**Tasks**:

1. **Implement `random_tensor(shape: List[Int], dtype: DType) -> ExTensor`**
   - Wrapper around `shared.testing.data_generators.random_tensor`
   - Default dtype: DType.float32
   - Uses deterministic seed for reproducibility

2. **Implement `sequential_tensor(shape: List[Int], dtype: DType) -> ExTensor`**
   - Create tensor with values 0, 1, 2, 3, ... (flattened, then reshaped)
   - Useful for testing index-dependent operations

3. **Implement `nan_tensor(shape: List[Int]) -> ExTensor`**
   - Create tensor filled with NaN values
   - **Note**: Defer to #2715 if NaN creation not yet supported

4. **Implement `inf_tensor(shape: List[Int]) -> ExTensor`**
   - Create tensor filled with infinity values
   - **Note**: Defer to #2715 if Inf creation not yet supported

5. **Implement `ones_like(tensor: ExTensor) -> ExTensor`**
   - Create tensor of ones matching input shape and dtype
   - Delegate to `shared.core.ones()` with shape from input

6. **Implement `zeros_like(tensor: ExTensor) -> ExTensor`**
   - Create tensor of zeros matching input shape and dtype
   - Delegate to `shared.core.zeros()` with shape from input

**Dependencies**: `shared.core.extensor`, `shared.testing.data_generators`

### Phase 3: Enhance Test Configuration (tests/conftest.mojo)

**File**: `/home/mvillmow/ml-odyssey/tests/conftest.mojo`
**Current State**: Basic constants defined (DEFAULT_RTOL, DEFAULT_ATOL), TODO for additional configuration

**Tasks**:

1. **Add multi-precision test constants**
   - `TEST_DTYPES` list for iteration across precision levels
   - Per-dtype tolerance mappings

2. **Implement `measure_time[func]()` function** (currently placeholder returning 0.0 at line 181)
   - Use `time.perf_counter_ns()` for nanosecond precision
   - Convert to milliseconds for return value

3. **Implement `measure_throughput[func](n_iterations)` function** (depends on measure_time)
   - Run function n_iterations times
   - Calculate operations per second

4. **Add TestFixtures tensor methods** (currently commented out at lines 69-93 with FIXME)
   - `small_tensor() -> ExTensor` - 3x3 tensor with known values
   - `random_tensor(rows, cols) -> ExTensor` - deterministic random tensor

5. **Add additional precision constants**
   - RTOL_FP32, RTOL_FP16, RTOL_BF16, RTOL_FP8

**Dependencies**: `time` module, `shared.core.extensor`

### Phase 4: Implement Integration Test Execution Logic (tests/shared/integration/__init__.mojo)

**File**: `/home/mvillmow/ml-odyssey/tests/shared/integration/__init__.mojo`
**Current State**: Empty `main()` with `pass` and TODO at line 16

**Tasks**:

1. **Implement test runner function**
   - `run_integration_tests() -> Int` - returns failure count
   - Run each test module in the integration directory

2. **Implement test result tracking**
   - `TestResult` struct with name, passed, duration_ms, error_message

3. **Implement test discovery helpers**
   - `run_test_safely()` - Wrap test in try/except, capture errors

4. **Update main() to run tests**
   - Execute all integration tests
   - Report pass/fail summary
   - Exit with error if any failures

**Dependencies**: Individual test modules in `tests/shared/integration/`

### Phase 5: Implement TOML Config Loading (tests/shared/integration/test_multi_precision_training.mojo)

**File**: `/home/mvillmow/ml-odyssey/tests/shared/integration/test_multi_precision_training.mojo`
**Current State**: Line 414 has TODO for TOML config loading in `test_training_with_toml_config()`

**Tasks**:

1. **Create TOML config parser struct**
   - `TrainingConfig` with fields: precision_mode, initial_scale, batch_size, learning_rate, epochs
   - `from_toml(path: String) -> Self` static method

2. **Implement `test_training_with_toml_config()`**
   - Load config from `configs/lenet5/emnist/fp16.toml`
   - Create PrecisionConfig from loaded values
   - Verify config values match TOML content

3. **Add TOML parsing helpers**
   - Simple line-by-line parser for key=value pairs
   - Handle [section] headers
   - Support string, int, float, bool types

4. **Add config validation tests**
   - Test loading all 4 precision configs (fp32, fp16, bf16, fp8)
   - Verify each creates correct PrecisionConfig

**Note**: TOML parsing in Mojo will require a simple custom implementation since native TOML support may not be available.

**Dependencies**: File I/O, `PrecisionConfig` from `shared.training.precision_config`

## Files to Modify/Create

| File | Action | Description |
|------|--------|-------------|
| `tests/helpers/utils.mojo` | Modify | Add 4 utility functions (print_tensor, tensor_summary, compare_tensors, benchmark) |
| `tests/helpers/fixtures.mojo` | Modify | Add 6 fixture functions (random_tensor, sequential_tensor, nan_tensor, inf_tensor, ones_like, zeros_like) |
| `tests/conftest.mojo` | Modify | Enhance with precision constants, implement measure_time/throughput, add TestFixtures methods |
| `tests/shared/integration/__init__.mojo` | Modify | Add test execution logic, TestResult struct, test runner |
| `tests/shared/integration/test_multi_precision_training.mojo` | Modify | Implement TOML config loading test (line 414) |
| `tests/helpers/__init__.mojo` | Modify | Export new utility and fixture functions |
| `tests/helpers/test_utils.mojo` | Create | Unit tests for utils.mojo functions |
| `tests/helpers/test_fixtures.mojo` | Create | Unit tests for fixtures.mojo functions |

## Testing Approach

### Unit Tests for New Functions

1. **tests/helpers/test_utils.mojo**
   - `test_print_tensor()` - verify output format
   - `test_tensor_summary()` - verify statistics calculation
   - `test_compare_tensors_identical()` - verify identical tensor comparison
   - `test_compare_tensors_different()` - verify difference detection
   - `test_benchmark()` - verify timing returns positive values

2. **tests/helpers/test_fixtures.mojo**
   - `test_random_tensor_shape()` - verify shape matches request
   - `test_sequential_tensor()` - verify values 0,1,2,3...
   - `test_ones_like()` - verify shape/dtype matching
   - `test_zeros_like()` - verify shape/dtype matching

### Integration Tests

1. **Verify existing tests still pass**
   ```bash
   pixi run mojo test tests/shared/integration/test_multi_precision_training.mojo
   ```

2. **Test TOML loading for all configs**
   - `configs/lenet5/emnist/fp32.toml`
   - `configs/lenet5/emnist/fp16.toml`
   - `configs/lenet5/emnist/bf16.toml`
   - `configs/lenet5/emnist/fp8.toml`

3. **Run full integration suite**
   ```bash
   pixi run mojo test tests/shared/integration/
   ```

### Validation Checklist

- [ ] All new functions have docstrings
- [ ] All new functions have unit tests
- [ ] No regressions in existing tests
- [ ] TOML config loading works for all 4 precision modes
- [ ] Integration test runner executes correctly
- [ ] Code follows Mojo guidelines (out self, mut self patterns)
- [ ] Pre-commit hooks pass (mojo format, markdownlint)

## Success Criteria

1. **Completeness**
   - All TODO(#2734) comments are resolved
   - All 5 target files have functional implementations
   - New functions are exported via `__init__.mojo`

2. **Quality**
   - Unit test coverage for all new functions
   - All tests pass locally and in CI
   - Documentation complete (docstrings, examples)

3. **Integration**
   - TOML config loading enables multi-precision training tests
   - Integration test runner can execute all test modules
   - New utilities integrate with existing `shared/testing/` infrastructure

4. **Backward Compatibility**
   - Existing test imports continue to work
   - No breaking changes to existing test code
   - `tests/shared/conftest.mojo` re-exports maintained

## Resource Usage

**Estimated Implementation Time**: 4-6 hours
- Phase 1 (utils.mojo): 1 hour
- Phase 2 (fixtures.mojo): 1 hour
- Phase 3 (conftest.mojo): 30 minutes
- Phase 4 (integration __init__.mojo): 1 hour
- Phase 5 (TOML loading): 1.5 hours
- Testing and documentation: 1 hour

**Tool Calls Used**: 17 (within budget of 50)
**Steps Used**: 12 (within budget of 50)

**Dependencies**:
- No external dependencies beyond existing `shared/` modules
- TOML parsing requires simple custom implementation

**Risk Factors**:
- TOML parsing complexity if native support unavailable (mitigation: simple line parser)
- NaN/Inf tensor creation may be blocked by ExTensor limitations (#2715) (mitigation: defer to #2715)
- Time measurement API availability in Mojo stdlib (mitigation: use available APIs or placeholder)
